File: sphone-api/Domain/Payments/PaymentItem.cs
using SPhone.Domain.Common;

namespace SPhone.Domain.Payments;

public sealed class PaymentItem : Entity<Guid>
{
    public PaymentScheduleId PaymentScheduleId { get; private set; }
    public int InstallmentNumber { get; private set; }
    public DateOnly DueDate { get; private set; }
    public decimal Amount { get; private set; }
    public PaymentStatus Status { get; private set; }
    public DateTime? PaidAt { get; private set; }

    // EF Core constructor
    private PaymentItem() : base(Guid.NewGuid())
    {
        PaymentScheduleId = null!;
    }

    internal PaymentItem(
        PaymentScheduleId scheduleId,
        int installmentNumber,
        DateOnly dueDate,
        decimal amount) : base(Guid.NewGuid())
    {
        PaymentScheduleId = scheduleId;
        InstallmentNumber = installmentNumber;
        DueDate = dueDate;
        Amount = amount;
        Status = PaymentStatus.Pending;
    }

    public void MarkAsPaid()
    {
        if (Status == PaymentStatus.Paid) throw new DomainException("Payment already paid.");
        Status = PaymentStatus.Paid;
        PaidAt = DateTime.UtcNow;
    }

    public void MarkAsOverdue()
    {
        if (Status != PaymentStatus.Pending) return;
        Status = PaymentStatus.Overdue;
    }
}

File: sphone-api/src/domain/payments/PaymentItem.ts
import { PaymentItemId } from './PaymentItemId'
import { PaymentStatus } from './PaymentStatus'

export class PaymentItem {
  private constructor(
    private readonly _id: PaymentItemId,
    private readonly _installmentNumber: number,
    private readonly _dueDate: Date,
    private readonly _amount: number,
    private _status: PaymentStatus,
    private _paidAt: Date | null,
  ) {}

  get id(): PaymentItemId { return this._id }
  get installmentNumber(): number { return this._installmentNumber }
  get dueDate(): Date { return this._dueDate }
  get amount(): number { return this._amount }
  get status(): PaymentStatus { return this._status }
  get paidAt(): Date | null { return this._paidAt }

  static create(params: {
    id: string
    installmentNumber: number
    dueDate: Date
    amount: number
  }): PaymentItem {
    return new PaymentItem(
      PaymentItemId.create(params.id),
      params.installmentNumber,
      params.dueDate,
      params.amount,
      PaymentStatus.Pending,
      null,
    )
  }

  static reconstitute(params: {
    id: string
    installmentNumber: number
    dueDate: Date
    amount: number
    status: number
    paidAt: Date | null
  }): PaymentItem {
    return new PaymentItem(
      PaymentItemId.create(params.id),
      params.installmentNumber,
      params.dueDate,
      params.amount,
      params.status as PaymentStatus,
      params.paidAt,
    )
  }
}
